Add Recordings section to the left panel

abey79 opened the pull request:
### What

Add Recordings section to the left panel.

Fixes #2298

Also:
- [x] created a brand new `ListItem` widget to be used across the UI
  - full-span highlight
  - icon support (`Icon` or closure)
  - hover button support
- [x] created a proper "panel_title_bar" widget in ReUi
- [x] removed the recordings submenu from the rerun menu
- [x] updated re_ui_example.rs to illustrate the "proper" panel hierarchy and demo the new widgets
- [x] `+` button in title to open new recording

Deferred to follow-up PRs:
- hierarchical display
- minus hover button to close the recording

<img width="1412" alt="image" src="https://github.com/rerun-io/rerun/assets/49431240/dd761936-bd5d-4161-b7c2-353652cf1897">


### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] ~~I have tested [demo.rerun.io](https://demo.rerun.io/pr/2938) (if applicable)~~ best tested on a native build and throwing lots of example at it, e.g. `for run in {1..4}; do python examples/python/dna/main.py --connect; done`

- [PR Build Summary](https://build.rerun.io/pr/2938)
- [Docs preview](https://rerun.io/preview/pr%3Aantoine%2Fnew-recording-ui-2298/docs)
- [Examples preview](https://rerun.io/preview/pr%3Aantoine%2Fnew-recording-ui-2298/examples)

martenbjork reviewed (approved):
Super! 🔥

emilk commented on crates/re_ui/src/list_item.rs:
```diff
@@ -0,0 +1,196 @@
+                .clone()
+                .into_galley(ui, Some(false), wrap_width, egui::TextStyle::Button);
+
+        let desired_size = (padding_extra + text.size() + egui::vec2(icon_extra, 0.0))
```
nit: the icon is to the left of the text

```suggestion
        let desired_size = (padding_extra + egui::vec2(icon_extra, 0.0) + text.size())
```

emilk commented on crates/re_ui/src/list_item.rs:
```diff
@@ -0,0 +1,196 @@
+
+            let min_pos = ui.painter().round_pos_to_pixels(egui::pos2(
+                rect.min.x.ceil(),
+                ((rect.min.y + rect.max.y - ReUi::small_icon_size().y) * 0.5).ceil(),
```
summing min and max looks like a bug until you read on

```suggestion
                (rect.center().y - 0.5 * ReUi::small_icon_size().y).ceil(),
```

emilk commented on crates/re_ui/src/list_item.rs:
```diff
@@ -0,0 +1,196 @@
+            bg_rect.extend_with_x(ui.clip_rect().left());
+            let background_frame = ui.painter().add(egui::Shape::Noop);
+
+            let min_pos = ui.painter().round_pos_to_pixels(egui::pos2(
+                rect.min.x.ceil(),
+                ((rect.min.y + rect.max.y - ReUi::small_icon_size().y) * 0.5).ceil(),
+            ));
```
what is the purpose of the `ceil` here? Do we want to align widgets on a `point` grid (even if there is sometimes a non-integral number of pixels per point)?

abey79 replied:
Not sure. Maybe @Wumpf knows something I don't. I copy/pasted that code from `ReUi::selectable_lable_with_icon()` without thinking much about it.

Original code in context:
https://github.com/rerun-io/rerun/blob/cdca75466aa6db3469163ba9e46822aacbe8007a/crates/re_ui/src/lib.rs#L638-L654

emilk replied:
I suggest we remove it from both places then (and check that it still looks good even when you zoom in)

abey79 replied:
OK

emilk commented on crates/re_ui/src/list_item.rs:
```diff
@@ -0,0 +1,196 @@
+            bg_rect.extend_with_x(ui.clip_rect().left());
+            let background_frame = ui.painter().add(egui::Shape::Noop);
+
+            let min_pos = ui.painter().round_pos_to_pixels(egui::pos2(
```
`icon_pos` seems a better name than `min_pos`

(positions are usually the top left)

abey79 replied:
I had decided against `icon_pos` because that's actually the "label pos" if we don't have an icon. Actually, the ambiguity is due to the `x` and `y` component being used separately at different places. I'll refactor that little bit of code to make it clearer.

emilk commented on crates/re_ui/src/list_item.rs:
```diff
@@ -0,0 +1,196 @@
+            // Draw text next to the icon.
+            let mut text_rect = rect;
+            text_rect.min.x = min_pos.x + icon_extra;
+            let text_pos = ui
+                .layout()
+                .align_size_within_rect(text.size(), text_rect)
```
This will follow the alignment of the current layout, but I think we should always align the label as left-center. You can use `Align2::LEFT_CENTER.align_size_within_rect` for that

abey79 replied:
Oh, that maybe explains some of the behaviour I've seen with `CollapsibleState`.

emilk commented on crates/re_ui/src/list_item.rs:
```diff
@@ -0,0 +1,196 @@
+        self
+    }
+
+    /// Provide a closure to display on-hover buttons on the right of the item.
```
You should document that the buttons will be layed out right-to-left

emilk commented on crates/re_viewer/src/app_state.rs:
```diff
@@ -146,12 +147,40 @@ impl AppState {
+                        ui.set_clip_rect(ui.max_rect());
+
+                        // TODO(ab): this might be promoted higher in the hierarchy once list item are
+                        //  used in the blueprint panel section
```
```suggestion
                        // used in the blueprint panel section.
```

abey79 replied:
hehe, the reason for that is:

<img width="614" alt="image" src="https://github.com/rerun-io/rerun/assets/49431240/29ec9477-5c88-4299-9d5f-ee915de600b3">

<img width="607" alt="image" src="https://github.com/rerun-io/rerun/assets/49431240/837c5df7-ab7a-416b-8a45-c32c7c3a0e62">

always, who cares 🤷🏼

emilk commented on crates/re_viewer/src/ui/recordings_panel.rs:
```diff
@@ -0,0 +1,104 @@
+#[cfg(not(target_arch = "wasm32"))]
+use re_ui::UICommandSender;
+
+/// Show the Recordings section of the left panel
```
it's good to describe what the code does: "Shows the currently open recordings in a selectable list"

emilk commented on crates/re_viewer/src/ui/recordings_panel.rs:
```diff
@@ -0,0 +1,104 @@
+        });
+}
+
+#[allow(clippy::blocks_in_if_conditions)]
```
I don't see an `if { … }`, but maybe I'm blind

emilk commented on crates/re_viewer/src/ui/recordings_panel.rs:
```diff
@@ -0,0 +1,104 @@
+
+    let active_recording = store_context.recording.map(|rec| rec.store_id());
+
+    let desc = format_description!(version = 2, "[hour]:[minute]:[second]");
```
we should make it clear(ish) that these are UTC times

```suggestion
    let desc = format_description!(version = 2, "[hour]:[minute]:[second]Z");
```

emilk commented on crates/re_viewer/src/ui/recordings_panel.rs:
```diff
@@ -0,0 +1,104 @@
+        return;
+    }
+
+    fn store_db_key(store_db: &StoreDb) -> (&str, Time) {
+        store_db.store_info().map_or(("", Time::default()), |info| {
+            (info.application_id.0.as_str(), info.started)
+        })
+    }
```
`Option` can be ordered

```suggestion
    fn store_db_key(store_db: &StoreDb) -> impl Ord {
        store_db.store_info().map(|info| {
            (info.application_id.0.as_str(), info.started)
        })
    }
```

abey79 replied:
Niiiice thanks.

emilk commented on crates/re_viewer/src/ui/recordings_panel.rs:
```diff
@@ -0,0 +1,104 @@
+use time::macros::format_description;
+
+#[cfg(not(target_arch = "wasm32"))]
+use re_ui::UICommandSender;
```
If this is only used from within `add_button_ui` I suggest adding the `use` statement to the contents that function (reducing the scope of the binding, and making it easier to move the function)

emilk commented on crates/re_viewer/src/ui/recordings_panel.rs:
```diff
@@ -0,0 +1,104 @@
+            ui,
+            "Recordings",
+            Some("These are the Recordings currently loaded in the Viewer"),
+            #[allow(unused_variables)]
+            |ui| {
+                #[cfg(not(target_arch = "wasm32"))]
+                add_button_ui(ctx, ui);
```
Alternative trick:
```suggestion
            |_ui| {
                #[cfg(not(target_arch = "wasm32"))]
                add_button_ui(ctx, _ui);
```

emilk commented on crates/re_viewer/src/ui/recordings_panel.rs:
```diff
@@ -0,0 +1,104 @@
+    if ctx
+        .re_ui
+        .small_icon_button(ui, &re_ui::icons::ADD)
+        .on_hover_text(re_ui::UICommand::Open.text_and_tooltip().1)
```
We should also mention the keyboard shortcut here! There is `format_shortcut_tooltip_suffix` you can use, and perhaps add a helper function for it:

```suggestion
        .on_hover_text(re_ui::UICommand::Open.text_with_shortcut(ui.ctx()))
```

abey79 replied:
Good point (I named it `tooltip_with_shortcut` though).

emilk commented on crates/re_viewer/src/ui/rerun_menu.rs:
```diff
@@ -3,14 +3,14 @@
     pub fn rerun_menu_button_ui(
         &mut self,
-        store_context: Option<&StoreContext<'_>>,
+        #[allow(unused_variables)] store_context: Option<&StoreContext<'_>>,
```
rm

emilk reviewed (approved):
Looks good! Nothing major to interject.

Feel free to merge once my comments has been addressed (or ignored as you see fit)

emilk commented:
> 

Yeah you are right, probably not a new bug, just more obvious now (which is a good thing!)

emilk commented:
> > When I use the `+` button to open new recording, the newly open recording is not automatically made the selected once
> 
> So does the "Open..." menu item then? It should be doing the same thing. Anyway, I'll take a look tomorrow.


Yeah you are right, probably not a new bug, just more obvious now (which is a good thing!)

abey79 commented on crates/re_ui/examples/re_ui_example.rs:
```diff
@@ -219,29 +224,94 @@ impl eframe::App for ExampleApp {
-                    .large_collapsing_header(ui, "Large Collapsing Header", true, |ui| {
+                ui.set_clip_rect(ui.max_rect());
+
+                // first section - no scroll area, so a global Frame can be used.
```
I meant a single outer wrapping `Frame` (actually using the `panel_content()` helper). I've improved the wording.

abey79 commented on crates/re_ui/examples/re_ui_example.rs:
```diff
@@ -219,29 +224,94 @@ impl eframe::App for ExampleApp {
+                // Second section. It's a list of `list_items`, so we need to remove the default
+                // spacing. Also, it uses a scroll area, so we must use several `Frame`s.
+                ui.scope(|ui| {
+                    ui.spacing_mut().item_spacing.y = 0.0;
```
I think so, though the styling of the title has already changed in the meantime and that is all part of a whole lot of things that will see significant styling adjustment and fine-tuning in the coming weeks (when Mårten is back from holidays).

abey79 commented on crates/re_ui/src/lib.rs:
```diff
@@ -444,6 +454,72 @@ impl ReUi {
+                ui.allocate_ui_with_layout(
+                    ui.available_size(),
+                    egui::Layout::right_to_left(egui::Align::Center),
+                    |ui| add_right_buttons(ui),
```
Well spotted. I'm surprised as well. Might be due to the fact that the closure type is generic over return type?

abey79 commented on crates/re_ui/src/list_item.rs:
```diff
@@ -0,0 +1,196 @@
+/// - selectable
+/// - full span highlighting
+/// - optional icon
+/// - optional on-hover buttons on the right
```
Pretty neat with PyCharm's "docstring preview" feature :)

<img width="794" alt="image" src="https://github.com/rerun-io/rerun/assets/49431240/d70b1cab-587a-4fb2-b8e1-eb5f9df1cad4">

abey79 commented:
@emilk Great review, thanks! (I learned a bunch on the way.)

I've addressed all the comments marked as resolved. I guess the `ceil()` is the remaining thing to clarify.

I also fixed the "loaded rrd isn't activated" bug. Maybe you'll want to take a look as I haven't touched that stuff much so far.

emilk commented on crates/re_ui/src/list_item.rs:
```diff
@@ -138,22 +150,22 @@ impl egui::Widget for ListItem<'_> {
+                let icon_pos = ui.painter().round_pos_to_pixels(egui::pos2(
+                    leftmost_x_pos,
+                    (rect.center().y - 0.5 * ReUi::small_icon_size().y).ceil(),
+                ));
```
let's remove the `.ceil()` calls
